### PR TITLE
feat: migrate to tanstack query v4

### PR DESCRIPTION
**Pull request by LetItRock**
### What change does this PR introduce? 

I was playing around with the app and noticed a couple of issues:
- when switching to production and back to development the "New template" button doesn't work
- in production the "New template" button is disabled but it's possible to click and navigate to the New Template page
- if you navigate to the New Template page in development and then switch to production the New Template page will be visible, disabled fields will show validation errors and is behaving weird

--

The first issue is related to the disabled `useTemplateFetcher` hook (react-query hook) which is using the `templateId` as the argument to fetch the template details.
The thing here was that when we do switch the environment we refetch all the queries with the `queryClient.refetchQueries` call, which in v3 has a bug that also forces the disabled queries to fetch. So it forced the `useTemplateFetcher` hook to call its fetch function which makes the request to `/v1/notification-templates/` and instead of returning only one template it returns an array of templates, which causes the issue.

--

In `@tanstack/react-query` v4 that issue is fixed, that's why I thought I will check their migration guide to see how hard to migrate it is. After reading the migration guide I understood that it's not that hard and we would probably do that anyway in the future, so I decided to give it a shot.

### Other information (Screenshots)

Issue screenshots:

https://user-images.githubusercontent.com/2607232/208324722-0ccab56e-3ee2-4d43-b7a8-3c1b3e93f143.mov

https://user-images.githubusercontent.com/2607232/208324732-54cadce4-5feb-427a-82ac-0c264358f1eb.mov

https://user-images.githubusercontent.com/2607232/208324741-1a6b6f4b-bf99-46cc-b1f1-c1d6ab4e1e14.mov

### CODE REVIEW

File: apps/web/cypress/tests/notification-editor/main-functionality.spec.ts [hunk 1]
```diff
@@ -310,4 +310,20 @@ describe('Workflow Editor - Main Functionality', function () {
...
+    cy.getByTestId('create-template-btn').should('be.disabled');
+  });
+
+  it('Should not allow to go to New Template page in Production', function () {
+    cy.visit('/templates/create');
+
+    cy.getByTestId('environment-switch').find(`input[value="Production"]`).click({ force: true });
+
+    cy.location('pathname').should('equal', `/templates`);
+  });
```
Comment (LetItRock) [inline apps/web/cypress/tests/notification-editor/main-functionality.spec.ts:328]: added some tests to cover the found bugs

File: apps/web/cypress/tests/notifications.spec.ts [hunk 2]
```diff
@@ -21,7 +21,7 @@ describe('Notification Templates Screen', function () {
       .getByTestId('active-status-label')
       .should('be.visible');
 
-    cy.getByTestId('create-template-btn').should('have.attr', 'href', '/templates/create');
+    cy.getByTestId('create-template-btn').should('not.be.disabled');
```
Comment (LetItRock) [inline apps/web/cypress/tests/notifications.spec.ts:24]: the New Template button was a link, and I've changed it to be a button

File: apps/web/cypress/tests/organization-switch.spec.ts [hunk 3]
```diff
@@ -20,7 +20,7 @@ describe('Organization Switch', function () {
 
       cy.getByTestId('organization-switch').focus();
 
-      cy.get('.mantine-Select-item').contains(capitalize(newOrg.name)).click();
+      cy.get('.mantine-Select-item').contains(capitalize(newOrg.name)).click({ force: true });
```
Comment (LetItRock) [inline apps/web/cypress/tests/organization-switch.spec.ts:23]: this e2e test was failing to me

File: apps/web/package.json [hunk 4]
```diff
@@ -48,6 +48,8 @@
     "@sentry/tracing": "^6.3.1",
     "@storybook/addon-docs": "^6.4.19",
     "@storybook/theming": "^6.4.19",
+    "@tanstack/react-query": "^4.20.4",
+    "@tanstack/react-query-devtools": "^4.20.4",
```
Comment (LetItRock) [inline apps/web/package.json:52]: tanstack react-query and devtools

File: apps/web/src/api/hooks/integrations/use-active-integrations.ts [hunk 5]
```diff
@@ -1,8 +1,9 @@
-import { useQuery } from 'react-query';
+import { useQuery } from '@tanstack/react-query';
+
 import { getActiveIntegrations } from '../../integration';
 
 export function useActiveIntegrations() {
-  const { data, isLoading, refetch } = useQuery('notificationsList', getActiveIntegrations);
+  const { data, isLoading, refetch } = useQuery(['notificationsList'], getActiveIntegrations);
```
Comment (LetItRock) [inline apps/web/src/api/hooks/integrations/use-active-integrations.ts:6]: I've run the codemods from the migration guide.
One of the changes they did was that query key should be now always an array

File: apps/web/src/components/activity/ExecutionDetailsModal.tsx [hunk 6]
```diff
@@ -24,10 +23,14 @@ export function ExecutionDetailsModal({
 }) {
   const theme = useMantineTheme();
   const [shouldRefetch, setShouldRefetch] = useState(true);
-  const { data: response, isLoading } = useQuery(['activity', notificationId], () => getNotification(notificationId), {
-    enabled: !!notificationId,
-    refetchInterval: shouldRefetch ? 1000 : false,
-  });
+  const { data: response, isInitialLoading } = useQuery(
```
Comment (LetItRock) [inline apps/web/src/components/activity/ExecutionDetailsModal.tsx:26]: another thing that was introduced `isInitialLoading`, which should be used when the query could be in the  [disabled state](https://tanstack.com/query/v4/docs/react/guides/migrating-to-react-query-4#disabled-queries)

File: apps/web/src/pages/templates/TemplatesListPage.tsx [hunk 7]
```diff
@@ -103,11 +108,14 @@ function NotificationList() {
...
-            </Button>
-          </Link>
+          <Button
+            disabled={readonly}
+            onClick={handleRedirectToCreateTemplate}
+            icon={<PlusCircle />}
+            data-test-id="create-template-btn"
+          >
+            New
+          </Button>
```
Comment (LetItRock) [inline apps/web/src/pages/templates/TemplatesListPage.tsx:118]: it was a link which was still clickable even the button was disabled, but now it's a regular button with a redirect logic on click

File: apps/web/src/pages/templates/editor/TemplateEditorPage.tsx [hunk 8]
```diff
@@ -90,12 +92,16 @@ export default function TemplateEditorPage() {
...
-          navigate('/templates/');
+          navigate('/templates');
         }
       }
     }
   }, [environment, template]);
 
+  if (environment && environment?.name === 'Production' && isCreateTemplatePage) {
+    navigate('/templates');
+  }
```
Comment (LetItRock) [inline apps/web/src/pages/templates/editor/TemplateEditorPage.tsx:103]: this if statement won't allow navigating to the create template page in the production

File: apps/web/src/App.tsx [hunk 9]
```diff
@@ -71,6 +72,7 @@ function App() {
     <HelmetProvider>
       <BrowserRouter basename={CONTEXT_PATH}>
         <QueryClientProvider client={queryClient}>
+          <ReactQueryDevtools />
```
Comment (scopsy) [inline apps/web/src/App.tsx:75]: Can we only show this in dev environment?

**Review by scopsy — APPROVED** [review]
Looks good! Always good idea to be uptodate without tools 🎉

File: apps/web/cypress/tests/organization-switch.spec.ts (hunk 3 above)
Comment (davidsoderberg) [reply apps/web/cypress/tests/organization-switch.spec.ts:23]: Is there a way to not need force here?

File: apps/web/src/App.tsx (hunk 9 above)
Comment (LetItRock) [reply apps/web/src/App.tsx:75]: sorry I forgot to push the last commit :D

File: apps/web/cypress/tests/organization-switch.spec.ts (hunk 3 above)
Comment (LetItRock) [reply apps/web/cypress/tests/organization-switch.spec.ts:23]: I haven't checked but saw a couple of times that we do force in other places. This test was failing to me locally but I'm wondering how it's passing on the next branch.

**Pull request merged**